modules/phase_field/src/userobjects/EulerAngleUpdater.C

```diff
@@ -0,0 +1,146 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "EulerAngleUpdater.h"
+#include "GrainTrackerInterface.h"
+#include "GrainForceAndTorqueInterface.h"
+#include "RotationTensor.h"
+
+template<>
+InputParameters validParams<EulerAngleUpdater>()
+{
+  InputParameters params = validParams<EulerAngleProvider>();
+  params.addClassDescription("Provide updated euler angles after rigid body rotation of the grains.");
+  params.addRequiredParam<UserObjectName>("grain_tracker_object", "The FeatureFloodCount UserObject to get values from.");
+  params.addRequiredParam<UserObjectName>("euler_angle_provider", "Name of Euler angle provider user object");
+  params.addRequiredParam<UserObjectName>("grain_torques_object", "Name of Euler angle provider user object");
+  params.addRequiredParam<VectorPostprocessorName>("grain_volumes", "The feature volume VectorPostprocessorValue.");
+  params.addParam<Real>("rotation_constant", 1.0, "Constant value characterizing grain rotation");
+  return params;
+}
+
+EulerAngleUpdater::EulerAngleUpdater(const InputParameters & params) :
+    EulerAngleProvider(params),
+    _grain_tracker(getUserObject<GrainTrackerInterface>("grain_tracker_object")),
+    _euler(getUserObject<EulerAngleProvider>("euler_angle_provider")),
+    _grain_torque(getUserObject<GrainForceAndTorqueInterface>("grain_torques_object")),
+    _grain_volumes(getVectorPostprocessorValue("grain_volumes", "feature_volumes")),
+    _mr(getParam<Real>("rotation_constant")),
+    _first_time(true)
+{
+}
+
+void
+EulerAngleUpdater::initialize()
+{
+  const auto grain_num = _grain_tracker.getTotalFeatureCount();
+
+  if (_first_time)
+  {
+    _angles.resize(grain_num);
+    _angles_old.resize(grain_num);
+    for (unsigned int i = 0; i < grain_num; ++i)
+      _angles[i] = _euler.getEulerAngles(i);
+  }
+
+  unsigned int angle_size = _angles.size();
+  for (unsigned int i = angle_size; i < grain_num; ++i)
+    _angles.push_back(_euler.getEulerAngles(i));
+
+  for (unsigned int i = 0; i < grain_num; ++i)
+  {
+    if (!_first_time)
+      if (!_fe_problem.converged())
+        _angles[i] = _angles_old[i];
+
+    RealGradient torque = _grain_torque.getTorqueValues()[i];
+
+    if (i <= angle_size)
+      _angles_old[i] = _angles[i];
+    else
+      _angles_old.push_back(_angles[i]);
+
+    RotationTensor R0(_angles_old[i]);
+    RealVectorValue torque_rotated = R0 * torque;
+    RealVectorValue omega = _mr / _grain_volumes[i] * torque_rotated;
+
+    RealVectorValue angle_change;
+    angle_change(0) = omega(2) * _dt;
```

Comment:
Please add a comment here, what are you trying to do here. Are you just building a rotation matrix for the torque effect?